### PR TITLE
[bugfix] Fix encoding issue for Windows

### LINKED ISSUE #13103: Invalid wallet path with Chinese characters in windows
Reported in https://github.com/bitcoin/bitcoin/pull/13100#issuecomment-384836977 by @ken2812221.

### LINKED ISSUE #13754: Windows crashes for -wallet=你好
Invalid characters for a wallet name, e.g. `bitcoin-qt -wallet=你好` used to result in an error like this

schermafbeelding 2018-07-24 om 17 46 37

But now it just crashes:

schermafbeelding 2018-07-24 om 17 43 26

I didn't do a `git bisect`, but it broke somewhere this year :-)

### PR DESCRIPTION
**Pull request by ken2812221**
Fix #13103 
Fix #13754

From https://github.com/bitcoin/bitcoin/pull/13107#issuecomment-385141831, @laanwj suggested that we should keep our internal strings to be always utf-8 encoded.

This PR adds two C++17 compatible macros: `u8string` and `u8path`. The reason that I use macros is that developers might not want to pass second parameter `utf8` everytime when they call `string` or create `path` objects.

I tried to find all `string` methods and convert them to `u8string` except for external api calls. Also tried to convert explicit or implicit `path` to `u8path`. Tell me if you find something that I missed.

Required: bitcoin-core/leveldb#18

See #13787 for travis run result

Caution: The user must change their config file to be utf-8 encoded if the file contains any non-ascii characters in it. Before 0.18, they are read as ansi-encoded characters on Windows.
--

### CODE REVIEW

**Comment by ken2812221** [comment]
>Can we use `path.imbued.locale` instead for this?

We still need to use ansi string when we call bdb and leveldb api, so I think that adding a new function is needed. Also, it needs to add `boost::locale` dependency in order to use boost locale generator.

**Comment by theuni** [comment]
I'm not really comfortable patching boost to make this work. Firstly because it would mean that only depends-builds can build for Windows, but also because I'm just not familiar with it.

Maybe start by upstreaming it and see where the discussion goes?

**Comment by ken2812221** [comment]
>Firstly because it would mean that only depends-builds can build for Windows

Is there any way to build Bitcoin Core on Windows without depends?

**Comment by achow101** [comment]
> Is there any way to build Bitcoin Core for Windows without depends-builds?

Supposedly there is a way to build with MSVC and Visual Studio which doesn't use the depends system.

File: src/util.cpp [hunk 1]
```diff
@@ -85,6 +85,8 @@ const int64_t nStartupTime = GetTime();
 const char * const BITCOIN_CONF_FILENAME = "bitcoin.conf";
 const char * const BITCOIN_PID_FILENAME = "bitcoind.pid";
 
+fs::detail::utf8_codecvt_facet g_utf8;
```
Comment (Empact) [inline src/util.cpp]: static?

File: src/util.h [hunk 2]
```diff
@@ -50,6 +50,11 @@ extern CTranslationInterface translationInterface;
 extern const char * const BITCOIN_CONF_FILENAME;
 extern const char * const BITCOIN_PID_FILENAME;
 
+extern fs::detail::utf8_codecvt_facet g_utf8;
+
+#define u8string() string(g_utf8)
+#define u8path(str) path(str, g_utf8)
```
Comment (Empact) [inline src/util.h]: Is it better to do these as methods in the namespace? Preprocessor use should be minimized IMO.

File: src/util.cpp (hunk 1 above)
Comment (ken2812221) [reply src/util.cpp]: `g_utf8` should be accessible globally.

File: src/util.h (hunk 2 above)
Comment (ken2812221) [reply src/util.h]: I just thought that pass the second parameter at every `.string()` and `path()` is pretty annoying and easy to forget. If most developers don't want to use macro function, I could do a scipted-diff.

Comment (Empact) [reply src/util.h]: I mean define these macros as functions in the fs namespace that call through as defined here.

File: src/util.cpp (hunk 1 above)
Comment (Empact) [reply src/util.cpp]: Looks like the only use is in util.h, if you implement those as methods in util.cpp, declared in util.h, you can limit variable access to here util.cpp. I may be wrong though! Reviewing on a cellphone. :P

File: src/util.h (hunk 2 above)
Comment (ken2812221) [reply src/util.h]: We could include `u8path` into fs namespace, however for `u8string` we couldn't since it's a member function.

Comment (Empact) [reply src/util.h]: I'll defer to others as to whether this is worthwhile, but absent other options, I would make a function ala `u8path` for string as well, and hide `g_utf8`

**Comment by ken2812221** [comment]
Rebased

**Review by ghost — APPROVED** [review]

File: src/bitcoin-cli.cpp [hunk 3]
```diff
@@ -422,6 +422,10 @@ static int CommandLineRPC(int argc, char *argv[])
             gArgs.ForceSetArg("-rpcpassword", rpcPass);
         }
         std::vector<std::string> args = std::vector<std::string>(&argv[1], &argv[argc]);
+#ifdef WIN32
+        for (std::string& arg : args)
+            arg = AnsiToUtf8(arg);
+#endif
```
Comment (ken2812221) [inline src/bitcoin-cli.cpp]: In order to solve this problem, there are three ways to solve this.
- The first is above, which might get `?` if the user enter any character that can't present by user's code page.
- The second is Use `GetCommandLineW` and `CommandLineToArgvW` to get command line options by `wchar_t`, which can present more Unicode characters.
- The third one is add wmain to get command line in `wchar_t` directly by "(w)main" function.

Which one is prefered?

**Comment by fanquake** [comment]
@theuni We probably want your thoughts here again.

For reference nothing has happened at the [upstream PR](https://github.com/boostorg/filesystem/pull/75), however that is less relevant now that we aren't patching Boost.

**Comment by Sjors** [comment]
Compared to master, this PR helps in at least the following scenario:
* Windows 10, system locale set to Simplified Chinese (see #13754 for how)
* launch QT from search bar or from command prompt: `bitcoin-qt -wallet=你好` 
* open console and do `getwalletinfo`

Before:
though

After:
schermafbeelding 2018-07-25 om 11 53 13

**Comment by ken2812221** [comment]
@Sjors Now you can use any walletname you want even if you are using English language setting.

File: src/bitcoin-cli.cpp (hunk 3 above)
Comment (ken2812221) [reply src/bitcoin-cli.cpp]: I choose the third way

**Comment by jonasschnelli** [comment]
Gitian Build: https://bitcoin.jonasschnelli.ch/build/715

**Comment by Sjors** [comment]
> Now you can use any walletname you want even if you are using English language setting

Thanks, that worked.

Can you change some of the functional tests to use unicode characters? I think `wallet_multiwallet.py` and `wallet_labels.py` would be good examples, since they use RPC arguments as well as filenames. I renamed label `e` in labels tests which worked fine on macOS. The multiwallet test was less happy when I renamed `w3`: `UnicodeEncodeError: 'ascii' codec can't encode characters in position 16-17: ordinal not in range(128)`. @jnewbery thoughts?

I also tested on macOS and that (still) works fine.
schermafbeelding 2018-07-26 om 13 35 04

**Comment by ken2812221** [comment]
Update: Dropped wmain, Use `GetCommandLineW` and `CommandLineToArgvW` instead.

**Comment by ken2812221** [comment]
Update: Now we can specify any characters in `-datadir`.

**Comment by ken2812221** [comment]
@Sjors Those needs upstream changes bitcoin-core/leveldb#18, you can see travis result on #13787.

**Comment by Empact** [comment]
If you call `CommandLineToArgvW` within `gArgs.ParseParameters` instead, you can avoid touching `src/bitcoind.cpp` and `src/bench/bench_bitcoin.cpp`, at least. There may be other opportunities for minimization.

**Comment by ken2812221** [comment]
>If you call CommandLineToArgvW within gArgs.ParseParameters

It would break the `ParseParameters` tests if we do that.

**Comment by ken2812221** [comment]
I'll seperate these into many PRs.

**Pull request closed without merging**